Migration from RxJava2 to Reactor

sdelamo opened the pull request:
Close: https://github.com/micronaut-projects/micronaut-core/issues/5554

Changes to Reactor PR: https://github.com/micronaut-projects/micronaut-reactor/pull/105

Changes to RxJava 3 https://github.com/micronaut-projects/micronaut-rxjava3/pull/114

This PR rewrites the internals of Micronaut to use Project Reactor as an implementation dependency. 

Micronaut RxJava 2 related code: instrumentation, converters and Rx Http Clients has been moved to https://github.com/micronaut-projects/micronaut-rxjava2

A custom operator has been written for the Lift and doAfterNext logic.

In general the following changes have been done to adjust the code:

## RxJava2 to Reactor 

### Flowable, Single, Maybe to Flux, Mono

`Flowable` is replaced with `Flux`

`Maybe` is replaced with `Mono`

`Single` is replaced with `Mono`

RxJava `Flowable.fromPublisher` to Reactor `Flux.from`

RxJava `Flowable::blockingFirst()` becomes `Flux::blockFirst()` in Reactor.

RxJava `Flowable::toList()` becomes `Flux::collectList()` in Reactor. 

RxJava `Flowable::blockingGet()` becomes `Flux::block()` in Reactor.

RxJava `Flowable::firstElement()` becomes `Flux:next()` in Reactor. 

`Flowable.fromCallable` becomes `Mono.fromCallable`

For usages of `Flowable::first` such as:

```java
Flowable.fromPublisher(...)
.first(Collections.emptyMap());
```

I have replaced it with: 

```java
Flux.from(...)
.defaultIfEmpty(Collections.emptyMap());
```
 
`Flowable::first` javadoc: 
 
 > Returns a Single that emits only the very first item emitted by this Flowable, or a default item if this Flowable completes without emitting anything.


 use `Flux.collectMap` to simplify code such as: 
 
```java
Map<String, Object> routeMap = new LinkedHashMap<>(routeList.size());

return Flowable.fromIterable(routeList)
    .collectInto(routeMap, (map, route) ->    map.put(getRouteKey(route), routeData.getData(route))
).toFlowable();
```

to: 

```java
return Flux.fromIterable(routeList)
                 .collectMap(this::getRouteKey, routeData::getData);\
```


### Emitter

`FlowableEmitter` becomes a `FluxSink`. And methods such as `onError`, `onNext` `onComplete` become `error`, `next` and `complete`. 

Replace usable of RxJava `SingleEmitter` with Project Reactor's `MonoSink`. 

`SingleEmitter::onSuccess` with `MonoSink::success`

I have replaced RxJava 2 `emitter.tryOnError(throwable)`

> Attempts to emit the specified Throwable error if the downstream hasn't cancelled the sequence or is otherwise terminated, returning false if the emission is not allowed to happen due to lifecycle restrictions.

with: 

```java
if (!emitter.isCanceled()) {
    emitter.error(throwable);
}
```


`emitter::request` becomes `emitter::requestedFromDownstream`

### Backpresssure

Instead of RxJava's `BackpressureStrategy`  it uses `FluxSink.OverflowStrategy.BUFFER`. 

### Delay

RxJava `.delay(1, TimeUnit.SECONDS);` to           `.delayElements(Duration.of(1, ChronoUnit.SECONDS));`

### Timeout

Rx Java `timeout(5, TimeUnit.SECONDS)` is replaced with Reactor's `.timeout(Duration.of(5, ChronoUnit.SECONDS))`

### Schedulers

Replaced the usage of `Schedulers.io()` with `Schedulers.boundedElastic`. 

RxJava:  

`Schedulers.from(ctx.channel().eventLoop()));`

Reactor:   

`Schedulers.fromExecutorService(ctx.channel().eventLoop())`

`Schedulers.from(executorService)` becomes `Schedulers.fromExecutorService(executorService))`

### ObserveOn, PublishOn SubscribeOn

RxJava `observeOn` `subscribeOn`, in reactor are `publishOn` and `subscribeOn`. 

### Cancellable

`Cancellable` is replaced with `Disposable`. `Cacellable::setCancelable`  `Disposable.onCancel`

### Completeable

`Mono<Void>` replaces the usages of RxJava `Completable`

sdelamo committed:
Migration from RxJava2 to Reactor

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
remove ignore

sdelamo committed:
Revert changes to static HttpClient.create

sdelamo committed:
add @param javadoc

sdelamo committed:
use Mono.fromCallable

sdelamo committed:
use collectList()

sdelamo committed:
don't ignore

jameskleeh commented on function-client/src/main/java/io/micronaut/function/client/http/HttpFunctionExecutor.java:
```diff
@@ -48,13 +48,13 @@
      * @param httpClient  The HTTP client
      */
-    public HttpFunctionExecutor(RxHttpClient httpClient) {
+    public HttpFunctionExecutor(ReactorHttpClient httpClient) {
```
Can this be changed to just HttpClient ?

sdelamo replied:
Changed to use `HttpClient`

jameskleeh commented:
Edit: nevermind - The generic being `ReactiveClient` and not `T` or something to that effect threw me off

sdelamo commented:
I am still working on this. @jameskleeh  I will ping you once I am done. I just push it as a Draft branch for my sanity.

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
Update StreamRequestSpec.groovy

sdelamo committed:
Replace tryOnError with !isCanceled => error

sdelamo committed:
use method reference

sdelamo committed:
remove comments

sdelamo committed:
replace doAfterNext with doOnNext

sdelamo committed:
replace first with defaultIfEmpty

sdelamo committed:
replace Compleable

sdelamo committed:
Use MonoSink instead of SingleEmitter

sdelamo committed:
observeOn to publishOn

sdelamo committed:
move todo before lift

sdelamo committed:
Use ReactorHttpClient instead of RxHttpClient

sdelamo committed:
move rxjava2 classes to its own module

sdelamo committed:
move rxjava2 to its own module

sdelamo committed:
bump up reactor to 2020.0.7

sdelamo committed:
use doOnEach instead of doAfterNext

sdelamo committed:
add reactor converter

sdelamo committed:
add reactor instrument

sdelamo committed:
use Mono.empty for Completeable

sdelamo committed:
use fromExecutorService

sdelamo committed:
rewrite as tryOnError

sdelamo committed:
Move RxJava2 binders to module

sdelamo committed:
refactor Split NettyBinderRegistrar

sdelamo committed:
replace TODOS for emitter.onSuccess

sdelamo committed:
Replace Flowable::first with defaultIfEmpty

sdelamo committed:
ObservableBodyBinder to rxjava2-http-server-netty

sdelamo committed:
remove unused imports

sdelamo committed:
Use HttpClient instead of ReactorHttpClient

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
fix tests

sdelamo committed:
temp concat the open bracket and the close bracket

sdelamo committed:
if exception onErrorResume empty

sdelamo committed:
use onErrorReturn Mono empty

sdelamo committed:
fix tests

sdelamo committed:
Use Mono.empty for Completeable

sdelamo committed:
remove unused import

sdelamo committed:
remove unused import

sdelamo committed:
create instrumentation package

sdelamo committed:
add back package-info.java of sse

sdelamo committed:
use lambda notation

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
use labmda syntax

sdelamo committed:
remove ununsed import

sdelamo committed:
removd doOnEach blocks

sdelamo committed:
replace forEach with doOnEach

sdelamo committed:
invoke repeat - 1

> numRepeat – the number of times to re-subscribe on onComplete (positive, or 0 for original sequence only)

sdelamo committed:
add courouting controller

sdelamo committed:
edit test

sdelamo committed:
replace .lift usage

sdelamo committed:
fix onErrorResume

sdelamo committed:
fix tests

sdelamo committed:
fix tests

sdelamo committed:
fix tests

sdelamo committed:
remove ignore if test seems to pass for JDK 11

sdelamo committed:
use Mono vs Single

sdelamo committed:
avoid bean pollution with spec.name

sdelamo committed:
revert changes to IdleTimeoutSpec

sdelamo committed:
ignore test

sdelamo committed:
add bridgeRxHttpClient

sdelamo committed:
fix test

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
test: delayElement syntax

sdelamo committed:
use next instead of firstElement

sdelamo committed:
test: correct test Mono.fromCallable syntax

sdelamo committed:
temporary ignore tests

sdelamo committed:
add missing license

sdelamo committed:
refactor: use generic for ReactiveHttpClientRegistry

sdelamo committed:
add logback.xml

sdelamo committed:
remove reactorbom

sdelamo committed:
remove testImplementation rxjava2

sdelamo committed:
remove unused import

sdelamo committed:
avoid bean pollution

sdelamo committed:
remove spec.name

sdelamo committed:
remove unused imports

sdelamo committed:
remove rxjava2 testImpl dep

sdelamo committed:
fix checkstyle

sdelamo committed:
avoid bean pollution

sdelamo committed:
refactor trailing bracket

sdelamo committed:
remove unused import

sdelamo committed:
refactor change getBeans to Publisher

sdelamo committed:
module rxjava2-websockets

sdelamo committed:
move test to rxjava

sdelamo committed:
rename remove Rx prefix

sdelamo commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:
```diff
@@ -1248,15 +1248,14 @@ public HttpContent apply(Object message) {
                                         // if the Publisher is returning JSON then in order for it to be valid JSON for each emitted element
                                         // we must wrap the JSON in array and delimit the emitted items
-                                        httpContentPublisher = Flowable.fromPublisher(httpContentPublisher)
-                                                .lift((FlowableOperator<HttpContent, HttpContent>) ArrayBracketSubscriber::new);
```
There is no `lift` method in Project Reactor.

sdelamo commented on http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java:
```diff
@@ -736,71 +736,59 @@ public void onComplete() {
-    public <I> Flowable<ByteBuffer<?>> dataStream(io.micronaut.http.HttpRequest<I> request) {
-        return Flowable.fromPublisher(resolveRequestURI(request))
-                .flatMap(buildDataStreamPublisher(request))
-                .doAfterNext(buffer -> {
```
Not sure how to map `doAfterNext`. I have tried with `doOnEach` but it is not the same. It breaks multiple tests.

sdelamo commented on http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java:
```diff
@@ -736,71 +736,59 @@ public void onComplete() {
-    public <I> Flowable<io.micronaut.http.HttpResponse<ByteBuffer<?>>> exchangeStream(io.micronaut.http.HttpRequest<I> request) {
-        return Flowable.fromPublisher(resolveRequestURI(request))
-                .flatMap(buildExchangeStreamPublisher(request))
-                .doAfterNext(res -> {
```
Not sure how to map `doAfterNext`. I have tried with `doOnEach` but it is not the same. It breaks multiple tests.

sdelamo committed:
add commented out blocks for doAfterNext

sdelamo committed:
annotate only the failing tests with @Ignore

sdelamo commented on test-suite/src/test/groovy/io/micronaut/tracing/jaeger/HttpTracingSpec.groovy:
```diff
@@ -81,6 +82,7 @@ class HttpTracingSpec extends Specification {
         }
     }
 
+    @Ignore(" https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test fails.

sdelamo commented on test-suite/src/test/groovy/io/micronaut/tracing/jaeger/HttpTracingSpec.groovy:
```diff
@@ -370,6 +373,7 @@ class HttpTracingSpec extends Specification {
         client.close()
     }
 
+    @Ignore(" https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test fails.

sdelamo commented on test-suite/src/test/groovy/io/micronaut/tracing/jaeger/HttpTracingSpec.groovy:
```diff
@@ -290,6 +292,7 @@ class HttpTracingSpec extends Specification {
         }
     }
 
+    @Ignore(" https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test fails.

sdelamo commented on test-suite/src/test/groovy/io/micronaut/tracing/jaeger/HttpTracingSpec.groovy:
```diff
@@ -508,6 +512,7 @@ class HttpTracingSpec extends Specification {
         client.close()
     }
 
+    @Ignore(" https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test fails.

sdelamo commented on test-suite/src/test/groovy/io/micronaut/tracing/jaeger/HttpTracingSpec.groovy:
```diff
@@ -695,6 +700,7 @@ class HttpTracingSpec extends Specification {
         client.close()
     }
 
+    @Ignore(" https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test fails.

sdelamo committed:
anntoate test wtih @Ignore with issue number

sdelamo commented on test-suite/src/test/groovy/io/micronaut/upload/UploadSpec.groovy:
```diff
@@ -300,20 +301,21 @@ class UploadSpec extends AbstractMicronautSpec {
         result == 'data.json: 16'
     }
 
+    @Ignore("https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test no longer throws an exception

sdelamo committed:
For StreamingClient use RxStreaming client at Factory

sdelamo committed:
add static method RxHttpClient::create

sdelamo committed:
annotated failing test wtih @Issue and number

sdelamo commented on http-client/src/test/groovy/io/micronaut/http/client/IdleTimeoutSpec.groovy:
```diff
@@ -17,14 +19,13 @@ import spock.util.concurrent.PollingConditions
 import java.lang.reflect.Field
 
 @Retry
+@Ignore(" https://github.com/micronaut-projects/micronaut-core/issues/5554")
```
This test fails

sdelamo committed:
remove duplicated mono

sdelamo commented on http-client/src/test/groovy/io/micronaut/http/client/OctetStreamSpec.groovy:
```diff
@@ -20,7 +22,8 @@ import java.nio.charset.StandardCharsets
 
 class OctetStreamSpec extends Specification {
```
This test is really Flaky

sdelamo committed:
extra Reactor binders to its own registrar

sdelamo committed:
fix checkstyle

sdelamo committed:
fix checkstyle

sdelamo committed:
move to different package

sdelamo committed:
rx http client factory

sdelamo committed:
add Jsonsubscriber

sdelamo committed:
correct Javadoc of MonoBodyBinder

sdelamo committed:
del RxHttpClient::create

sdelamo committed:
checkstyle: fix javadoc

sdelamo committed:
fix http tracing test

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
Create doOnEachAggregators

sdelamo committed:
decouple from ReactorClient

sdelamo committed:
generic DoAfterNextConsumer

sdelamo committed:
fix: kotlin compilation

sdelamo committed:
wip

sdelamo committed:
Split Bridged Clients

sdelamo committed:
add back Reactor converter

sdelamo committed:
add back json subscriber and doAfterNext (#5659)

* add back json subscriber and doAfterNext

* Fix request context propagation. Change tests to use .toBlocking() where appropriate

Co-authored-by: jameskleeh <james.kleeh@gmail.com>

jameskleeh committed:
Copy upload spec to rxjava-netty prior to conversion

jameskleeh committed:
Refactor existing test to reflect behavior of mono

jameskleeh committed:
Fix checkstyle

jameskleeh committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
remove unused ignore import

sdelamo committed:
change javadoc

sdelamo committed:
rename from rxHttpClient to reactiveHttpClient

sdelamo committed:
annotate return type as @NonNull

sdelamo committed:
test: add CompleteableSpec

sdelamo commented on http-client/src/test/groovy/io/micronaut/http/client/HttpGetSpec.groovy:
```diff
@@ -426,18 +416,6 @@ class HttpGetSpec extends Specification {
         client.close()
     }
 
-    void "test completable returns 200"() {
```
I've moved this test to `rxjava2-http-client`

sdelamo committed:
extract MonoPost test

sdelamo committed:
test: maybe and single post

sdelamo committed:
remove unecessary requires

sdelamo committed:
fix: convert to publisher not to flux

sdelamo committed:
don't include reactor converters in runtime

- Replace methods returning Mono with @SingleResult Publisher
- Replace mehtods returning Flux to Publisher

sdelamo committed:
Improve docs

sdelamo committed:
test: reanme to echo-flow instaed of echo-publisher

sdelamo committed:
rename from rx to reactive

sdelamo committed:
remove reactor and rxjava2 modules

sdelamo committed:
add a snippet to what's new about project reactor

sdelamo committed:
test: MDCSpec Fix doFinally

sdelamo committed:
javadoc: set version to 3.0 and author to sergio

sdelamo commented on http-client-core/src/main/java/io/micronaut/http/client/RxProxyHttpClient.java:
```diff
@@ -1,31 +0,0 @@
- * @author graemerocher
- * @since 2.0.0
- */
-public interface RxProxyHttpClient extends ProxyHttpClient {
```
Moved to its own module

sdelamo commented on http-client-core/src/main/java/io/micronaut/http/client/RxStreamingHttpClient.java:
```diff
@@ -1,75 +0,0 @@
- * @author Graeme Rocher
- * @since 1.0
- */
-public interface RxStreamingHttpClient extends StreamingHttpClient, RxHttpClient {
```
Moved to its own module

sdelamo commented on http-client-core/src/main/java/io/micronaut/http/client/rxjava2/RxReactiveClientResultTransformer.java:
```diff
@@ -1,59 +0,0 @@
-@Requires(classes = Flowable.class)
-@Internal
-@Indexed(ReactiveClientResultTransformer.class)
-public class RxReactiveClientResultTransformer implements ReactiveClientResultTransformer {
```
Moved to its own module.

sdelamo commented on http-client-core/src/main/java/io/micronaut/http/client/sse/RxSseClient.java:
```diff
@@ -1,52 +0,0 @@
- * @author Graeme Rocher
- * @since 1.0
- */
-public interface RxSseClient extends SseClient {
```
Moved to its own module

sdelamo committed:
don't use Flux in protected method signatures

sdelamo committed:
checkstyle: sentence must end with period

sdelamo committed:
checkstyle: remove unused import

sdelamo committed:
rename to NettyWebSocketSession

sdelamo committed:
change Mono to Publisher SingleResult

sdelamo committed:
use String instead of def

sdelamo commented on http-client/src/test/groovy/io/micronaut/http/client/rxjava2/RxHttpPostSpec.groovy:
```diff
@@ -116,43 +120,15 @@ class RxHttpPostSpec extends Specification {
         book.title == "The Stand"
     }
 
-    void "test reactive single post retrieve request with JSON"() {
```
Moved to the RxJava project

sdelamo committed:
change to publisher

sdelamo committed:
change Flux to PUblisher

sdelamo committed:
remove RxJava from javadoc

sdelamo committed:
remove comments

sdelamo committed:
remove variable

sdelamo committed:
replace Mono and Flux with Publisher

sdelamo committed:
Change Mono to Publisher SingleResult

sdelamo committed:
rename to DefaultBeanDefinitionDataCollector

sdelamo committed:
change Flux to Publisher in protected methods

sdelamo committed:
Rename to DefaultRouteDataCollector

sdelamo committed:
replace Publisher with SingleResult

sdelamo committed:
document class rename

sdelamo committed:
use HttpClient instead of ReactorHttpClient

sdelamo committed:
replace Mono by Publisher SingleResult

sdelamo committed:
remove flux()

sdelamo committed:
remove empty folder

sdelamo committed:
change from Flux to Publisher

sdelamo committed:
change copy

sdelamo committed:
change from Mono to Publisher

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
make WebSocketStateBinderRegistry final

sdelamo committed:
remove empty blank line

sdelamo committed:
add testImplemenation to inject-groovy/build.gradle

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
add import

sdelamo committed:
don't use reactor bom

sdelamo committed:
reduce IdleTimeoutSpec flakyness

sdelamo committed:
test: remove @Ignore

sdelamo committed:
add static method to create websocket client

sdelamo committed:
rename to DefaultNettyHttpClientRegistry

sdelamo committed:
checkstyle fix

sdelamo committed:
add resolveClient methods to DefaultNetyClientRegistry

sdelamo committed:
add resolveProxyClient

sdelamo committed:
split client registries

jameskleeh committed:
Split out client factories

jameskleeh committed:
Rename tests. Add test to assert static creation works

jameskleeh committed:
Add javadoc and licenses

jameskleeh committed:
Refactor factory resolvers. Fix imports

jameskleeh committed:
Merge branch '3.0.x' into rxjava2-reactor

jameskleeh committed:
Fix test

jameskleeh committed:
Fix typo

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/sse/SseClient.java:
```diff
@@ -95,4 +99,28 @@
+     * @param url The base URL
+     * @return The client
+     */
+    static SseClient create(@Nullable URL url) {
+        return SseClientFactoryResolver.getFactory().createSseClient(url);
+    }
```
It should be documented with each `create` method that the returned client is not subject to dependency injection and that the creator is responsible for closing the client to avoid leaking connections

sdelamo replied:
Done via: https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/cc6aa8d2d0ec0b58be0a3db3a5de8f2c40fa780c

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/sse/SseClientFactoryResolver.java:
```diff
@@ -0,0 +1,52 @@
+ * @since 3.0.0
+ */
+@Internal
+class SseClientFactoryResolver {
```
Make `final`

sdelamo replied:
I've made the `ClientFactoryResolver` classes final via: https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/209af630430e8c6cd1df869e2fba7e3355698dc5

graemerocher commented on http-client-core/src/main/java/io/micronaut/http/client/sse/SseClientRegistry.java:
```diff
@@ -0,0 +1,72 @@
+    E resolveSseClient(@Nullable InjectionPoint<?> injectionPoint,
+                       @Nullable LoadBalancer loadBalancer,
+                       @Nullable HttpClientConfiguration configuration,
+                       BeanContext beanContext);
```
Add `@NonNull` or `@Nullable`

sdelamo replied:
I have added `@NonNull` to parameter `BeanContext beanContext` via : https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/5e182ac6040f160c126d805d1dcc6806ad9ef058

graemerocher commented on http-client/src/main/java/io/micronaut/http/client/netty/DefaultNettyHttpClientRegistry.java:
```diff
@@ -36,17 +36,13 @@
-import io.micronaut.http.client.RxHttpClient;
-import io.micronaut.http.client.RxHttpClientRegistry;
-import io.micronaut.http.client.RxStreamingHttpClient;
+import io.micronaut.http.client.*;
```
Thought we agreed not to use star imports?

sdelamo replied:
Done via https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/6a64732ebbb717a8e1ad32388b53c31c01e70d04

graemerocher commented on http-client/src/main/java/io/micronaut/http/client/netty/MicronautFlux.java:
```diff
@@ -0,0 +1,109 @@
+        }
+
+        @Override
+        public void subscribe(CoreSubscriber<? super T> actual) {
+            source.subscribe(new CoreSubscriber<T>() {
+                @Override
+                public Context currentContext() {
+                    return actual.currentContext();
+                }
+
+                @Override
+                public void onSubscribe(Subscription s) {
+                    actual.onSubscribe(s);
+                }
+
+                @Override
+                public void onNext(T t) {
+                    actual.onNext(t);
+                    afterNext.accept(t);
+                }
+
+                @Override
+                public void onError(Throwable t) {
+                    actual.onError(t);
+                }
+
+                @Override
+                public void onComplete() {
+                    actual.onComplete();
+                }
+            });
+        }
```
I believe you need to override the default method `currentContext` in `CoreSubscriber` otherwise this will lead to context loss. See https://bsideup.github.io/posts/daily_reactive/context_loss_detection/

sdelamo replied:
Not sure I understand. We are already overriding `currentContext` in this  snippet.

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/RoutingInBoundHandler.java:
```diff
@@ -1191,64 +1197,65 @@ private boolean isSingle(RouteMatch<?> finalRoute, Class<?> bodyClass) {
-                                            Object nativeBuffer = byteBuffer.asNativeBuffer();
-                                            if (nativeBuffer instanceof ByteBuf) {
-                                                httpContent = new DefaultHttpContent((ByteBuf) nativeBuffer);
+                                    Publisher<HttpContent> httpContentPublisher = Publishers.map(bodyPublisher, new Function<Object, HttpContent>() {
```
I think we may need to review the use of `Publisher.map` as it may result in context loss

graemerocher commented on http-server-netty/src/main/java/io/micronaut/http/server/netty/jackson/JsonViewServerFilter.java:
```diff
@@ -73,24 +74,25 @@ public JsonViewServerFilter(
                     if (Publishers.isConvertibleToPublisher(body)) {
-                        response.body(Publishers.convertPublisher(body, Flowable.class)
+                        Publisher<?> pub = Publishers.convertPublisher(body, Publisher.class);
+                        response.body(Flux.from(pub)
```
what if `pub` is already a `Flux`?

sdelamo replied:
I think it is fine due to the implementation of Flux: 

```java
public static <T> Flux<T> from(Publisher<? extends T> source) {
		//duplicated in wrap, but necessary to detect early and thus avoid applying assembly
		if (source instanceof Flux) {
			@SuppressWarnings("unchecked")
			Flux<T> casted = (Flux<T>) source;
			return casted;
		}
...
```

graemerocher commented on management/src/main/java/io/micronaut/management/endpoint/beans/BeansEndpoint.java:
```diff
@@ -52,13 +52,12 @@ public BeansEndpoint(BeanContext beanContext, BeanDefinitionDataCollector beanDe
      */
     @Read
-    public Single getBeans() {
+    public Publisher<?> getBeans() {
```
This needs to be annotated with `@SingleResult` otherwise it will now return transfer coding chunked which would be a change in behaviour

sdelamo replied:
done via https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/996aa4df47e0ea70c30abb1e5a619b22bc1096f0

graemerocher commented on management/src/main/java/io/micronaut/management/health/aggregator/RxJavaHealthAggregator.java:
```diff
@@ -23,10 +23,10 @@
 import io.micronaut.management.health.indicator.HealthIndicator;
```
Class needs renaming since it doesn't use RxJava anymore

sdelamo replied:
Renamed to `DefaultHealthAggregator` https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/8a765fcd35643f0ac6aa9b3dd995ca297606594b

sdelamo committed:
Merge branch '3.0.x' into rxjava2-reactor

sdelamo committed:
rename RxJavaHealthAggregator to DefaultHealthAggregator

sdelamo committed:
annotate getBeans with @SingleResult

sdelamo committed:
checkstyle: end first sentence with period

sdelamo committed:
remove wildcard import

sdelamo committed:
add @NonNull to BeanContext

sdelamo committed:
make ClientFactoryResolver classes final

sdelamo committed:
javadoc: indicate Client::create is not di aware

sdelamo committed:
add global spock extension to track context loss

sdelamo commented:
I am trying to add a Spock global extension to help with the debugging of context loss: https://github.com/micronaut-projects/micronaut-core/pull/5583/commits/c0194a58980b81ecef9909f27b56a45734f12b14

jameskleeh committed:
Merge remote-tracking branch 'origin/3.0.x' into rxjava2-reactor

jameskleeh committed:
Fix imports and javadoc

ilopmar commented:
:tada:
